tools/eval/main.go

```diff
@@ -0,0 +1,433 @@
+	}
+	// NOTE: Search within the top-level directory, whatever its name.
+	basename := filepath.SplitList(zr.File[0].Name)[0]
+	fullpath := filepath.Join(basename, path)
+	f, err := zr.Open(fullpath)
+	if err != nil {
+		return "", errors.Wrapf(err, "opening %q from zip", fullpath)
+	}
+	defer f.Close()
+	b, err := io.ReadAll(f)
+	if err != nil {
+		return "", errors.Wrap(err, "reading file contents")
+	}
+	return string(b), nil
+}
+
+func generateRecovery(ctx context.Context, model *genai.GenerativeModel, cache rebuild.AssetStore, r rundex.Rebuild, commands []string) (*llm.ScriptResponse, error) {
+	outcome := classifyOutcome(r)
+	s, err := r.Strategy.Strategy()
+	if err != nil {
+		return nil, errors.New("no strategy for rebuild")
+	}
+	switch outcome {
+	case OutcomeCompareMismatch:
+		packageJSON, err := fromSourceZip(ctx, cache, r.Target(), filepath.Join(location(s).Dir, "package.json"))
+		if err != nil {
+			return nil, errors.Wrap(err, "reading package.json")
+		}
+		return llm.InferNPMBuild(ctx, *model, packageJSON)
+	case OutcomeFailedRunBuild:
+		packageJSON, err := fromSourceZip(ctx, cache, r.Target(), filepath.Join(location(s).Dir, "package.json"))
+		if err != nil {
+			return nil, errors.Wrap(err, "reading package.json")
+		}
+		reader, err := cache.Reader(ctx, rebuild.Asset{Target: r.Target(), Type: rebuild.DebugLogsAsset})
+		if err != nil {
+			return nil, errors.Wrap(err, "creating build log reader")
+		}
+		logBytes, err := io.ReadAll(reader)
+		if err != nil {
+			return nil, errors.Wrap(err, "reading build log")
+		}
+		buildLog := string(logBytes)
+		return llm.FixNPMBreakage(ctx, *model, strings.Join(commands, "\n"), packageJSON, buildLog)
+	default:
+		return nil, errors.Errorf("unhandled outcome '%s'", outcome)
+	}
+}
+
+func validateFlags() error {
+	if *project == "" || *modelProject == "" || *baseModel == "" || *apiURI == "" || *debugStorage == "" || *runFlag == "" {
+		return fmt.Errorf("all flags except -dry-run are required")
+	}
+	switch *baseModel {
+	case "pro":
+		*baseModel = llm.GeminiPro
+	case "flash":
+		*baseModel = llm.GeminiFlash
+	case llm.GeminiPro, llm.GeminiFlash:
+	default:
+		return fmt.Errorf("model must be one of: %s, %s", llm.GeminiPro, llm.GeminiFlash)
+	}
+	return nil
+}
+
+func must[T any](t T, err error) T {
+	if err != nil {
+		panic(err)
+	}
+	return t
+}
+
+func main() {
+	flag.Parse()
+	if err := validateFlags(); err != nil {
+		log.Printf("Parsing flags: %v", err)
+		flag.Usage()
+		os.Exit(1)
+	}
+
+	ctx := context.Background()
+
+	// Initialize dependencies
+	client := must(genai.NewClient(ctx, *modelProject, "us-central1"))
+	model := client.GenerativeModel(*baseModel)
+	model = llm.WithSystemPrompt(*model, llm.NPMSystemPrompt)
+	apiURL := must(url.Parse(*apiURI))
+	idclient := must(oauth.AuthorizedUserIDClient(ctx))
+	dex := must(rundex.NewFirestore(ctx, *project))
+	debug := must(rebuild.NewGCSStore(context.WithValue(ctx, rebuild.RunID, *runFlag), *debugStorage))
```

Comment:
optional: Maybe worth putting the RunID in the main-level ctx? IIUC that would match the ctx scope and the pertinance of the RunID.

On the other hand, I can see the argument for using WithValue as late as possible though, it minimizes the amount of code that can be dependent on the RunID value (mitigating the downsides of global state).